datatypes/src/primitives/spatial_partition.rs

```diff
@@ -0,0 +1,221 @@
+            self.lower_right_coordinate.y,
+            self.upper_left_coordinate.y,
+        );
+
+        overlap_x && overlap_y
+    }
+
+    /// Returns the intersection (common area) of the partition with `other` if there is any
+    pub fn intersection(&self, other: &Self) -> Option<Self> {
+        if self.intersects(other) {
+            let ul_x = f64::max(self.upper_left_coordinate.x, other.upper_left_coordinate.x);
+            let ul_y = f64::min(self.upper_left_coordinate.y, other.upper_left_coordinate.y);
+            let lr_x = f64::min(
+                self.lower_right_coordinate.x,
+                other.lower_right_coordinate.x,
+            );
+            let lr_y = f64::max(
+                self.lower_right_coordinate.y,
+                other.lower_right_coordinate.y,
+            );
+
+            Some(Self::new_unchecked(
+                (ul_x, ul_y).into(),
+                (lr_x, lr_y).into(),
+            ))
+        } else {
+            None
+        }
+    }
+
+    /// Return true if the partition contains the `other`
+    pub fn contains(&self, other: &Self) -> bool {
+        self.contains_x(other) && self.contains_y(other)
+    }
+
+    fn contains_x(&self, other: &Self) -> bool {
+        crate::util::ranges::value_in_range(
+            other.upper_left_coordinate.x,
+            self.upper_left_coordinate.x,
+            self.lower_right_coordinate.x,
+        ) && crate::util::ranges::value_in_range(
+            other.lower_right_coordinate.x,
+            self.upper_left_coordinate.x,
+            self.lower_right_coordinate.x,
+        )
+    }
+
+    fn contains_y(&self, other: &Self) -> bool {
+        crate::util::ranges::value_in_range(
+            other.lower_right_coordinate.y,
+            self.lower_right_coordinate.y,
+            self.upper_left_coordinate.y,
+        ) && crate::util::ranges::value_in_range(
+            other.upper_left_coordinate.y,
+            self.lower_right_coordinate.y,
+            self.upper_left_coordinate.y,
+        )
+    }
+}
+
+pub trait SpatialPartitioned {
+    fn spatial_partition(&self) -> SpatialPartition;
+}
+
+impl BoxShaped for SpatialPartition {
+    fn upper_left(&self) -> Coordinate2D {
+        self.upper_left_coordinate
+    }
+
+    fn lower_right(&self) -> Coordinate2D {
+        self.lower_right_coordinate
+    }
+
+    fn upper_right(&self) -> Coordinate2D {
+        Coordinate2D {
+            x: self.lower_right_coordinate.x,
+            y: self.upper_left_coordinate.y,
+        }
+    }
+
+    fn lower_left(&self) -> Coordinate2D {
+        Coordinate2D {
+            x: self.upper_left_coordinate.x,
+            y: self.lower_right_coordinate.y,
+        }
+    }
+
+    fn size_x(&self) -> f64 {
+        self.lower_right_coordinate.x - self.upper_left_coordinate.x
+    }
+
+    fn size_y(&self) -> f64 {
+        self.upper_left_coordinate.y - self.lower_right_coordinate.y
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I guess you should also test the intersection, contains, etc.

Reply:
319239ad65edb889ea3181ed6e30f0a9674a5dfd